caikit_tgis_backend/managed_tgis_subprocess.py

```diff
@@ -0,0 +1,371 @@
+            self._wrapped_client = GenerationServiceStubWrapper(self)
+            return self._wrapped_client
+
+        log.warning("<MTS7717800W>", "get_client called with _tgis_client set to None")
+        return None
+
+    def launch(self):
+        """Launch the subprocess or restart it if it exists"""
+        with self._mutex:
+            self._launch()
+
+    def wait_until_ready(self, timeout=None) -> bool:
+        start_t = time.time()
+
+        # wait for the boot monitoring thread if we have one
+        if self._bootup_thread:
+            self._bootup_thread.join(timeout)
+
+        while True:
+            log.debug("wait_until_ready saw state [%s]", self._tgis_state)
+            if self._tgis_state == _TGISState.READY:
+                return True
+
+            if self._tgis_state == _TGISState.STOPPED:
+                # raise the bootup exception if we captured one
+                if self._bootup_exc:
+                    raise self._bootup_exc
+                return False
+
+            if time.time() - start_t >= timeout:
+                return False
+
+            time.sleep(self._bootup_poll_delay)
+
+    def terminate(self):
+        """Terminate the subprocess, wait for it to exit"""
+        with self._mutex:
+            self._ensure_terminated()
+
+    def _ensure_terminated(self):
+        """Terminate the subprocess if it exists
+
+        _mutex should be locked before calling this
+        """
+        if self._tgis_state != _TGISState.STOPPED:
+            if self._tgis_proc:
+                self._tgis_proc.terminate()
+                self._tgis_proc = None
+                # disown the bootup thread (but keep the exception if one was captured)
+                self._bootup_thread = None
+            log.debug("_ensure_terminated setting the state to STOPPED")
+            self._tgis_state = _TGISState.STOPPED
+
+    def _launch(self):
+        """Launch the subprocess or restart if it exists
+
+        _mutex should be locked before calling
+        """
+        self._ensure_terminated()
+
+        # Launch TGIS
+        launch_cmd = " ".join(
+            [
+                "text-generation-launcher",
+                "--num-shard 1",
+                f"--model-name {self._model_path}",
+                f"--port {self._http_port}",
+            ]
+        )
+        log.debug2("Launching TGIS with command: [%s]", launch_cmd)
+        env = os.environ.copy()
+        env["GRPC_PORT"] = str(self._grpc_port)
+        # Long running process
+        # pylint: disable=consider-using-with
+        self._tgis_proc = subprocess.Popen(shlex.split(launch_cmd), env=env)
+        log.debug("_launch setting state to BOOTING")
+        self._tgis_state = _TGISState.BOOTING
+        # launch background thread to monitor the boot sequence
+        # save a handle to it to record bootup errors
+        self._bootup_thread = threading.Thread(target=self._monitor_bootup, daemon=True)
```

Comment:
Interesting! I've never actually used `daemon` like this, but reading [the docs](https://docs.python.org/3/library/threading.html#thread-objects) this makes perfect sense